doc/sphinx/proof-engine/tactics.rst

```diff
@@ -650,180 +649,339 @@ Applying theorems
+      transforming your goal with the :tacn:`change` or :tacn:`pattern`
+      tactics.
 
-   .. tacv:: simple notypeclasses refine @term
-      :name: simple notypeclasses refine
+   .. exn:: Statement without assumptions.
 
-      This tactic behaves like the combination of :tacn:`simple refine` and
-      :tacn:`notypeclasses refine`: it performs type checking without resolution of
-      typeclasses, does not perform beta reductions or shelve the subgoals.
+      This happens if the type of :token:`term` has no non-dependent premise.
 
-   :opt:`Debug` ``"unification"`` enables printing traces of
-   unification steps used during elaboration/typechecking and the
-   :tacn:`refine` tactic. ``"ho-unification"`` prints information
-   about higher order heuristics.
+   .. exn:: Unable to apply.
 
-.. tacn:: apply @term
-   :name: apply
-
-   This tactic applies to any goal. The argument term is a term well-formed in
-   the local context. The tactic :tacn:`apply` tries to match the current goal
-   against the conclusion of the type of :token:`term`. If it succeeds, then
-   the tactic returns as many subgoals as the number of non-dependent premises
-   of the type of term. If the conclusion of the type of :token:`term` does
-   not match the goal *and* the conclusion is an inductive type isomorphic to
-   a tuple type, then each component of the tuple is recursively matched to
-   the goal in the left-to-right order.
-
-   The tactic :tacn:`apply` relies on first-order unification with dependent
-   types unless the conclusion of the type of :token:`term` is of the form
-   :n:`P (t__1 ... t__n)` with ``P`` to be instantiated. In the latter case,
-   the behavior depends on the form of the goal. If the goal is of the form
-   :n:`(fun x => Q) u__1 ... u__n` and the :n:`t__i` and :n:`u__i` unify,
-   then :g:`P` is taken to be :g:`(fun x => Q)`. Otherwise, :tacn:`apply`
-   tries to define :g:`P` by abstracting over :g:`t_1 ... t__n` in the goal.
-   See :tacn:`pattern` to transform the goal so that it
-   gets the form :n:`(fun x => Q) u__1 ... u__n`.
-
-   .. exn:: Unable to unify @term with @term.
-
-      The :tacn:`apply` tactic failed to match the conclusion of :token:`term`
-      and the current goal. You can help the :tacn:`apply` tactic by
-      transforming your goal with the :tacn:`change` or :tacn:`pattern`
-      tactics.
+      This happens if the conclusion of :token:`ident` does not match any of
+      the non-dependent premises of the type of :token:`term`.
 
    .. exn:: Unable to find an instance for the variables {+ @ident}.
 
-      This occurs when some instantiations of the premises of :token:`term` are not deducible
+      This occurs when some instantiations of the premises of :token:`one_term` are not deducible
       from the unification. This is the case, for instance, when you want to apply a
-      transitivity property. In this case, you have to use one of the variants below:
+      transitivity property.
```

Comment:
We should add the advice to use the `with @bindings` clause or the `eapply` tactic in this case.